elements/lisk-utils/src/math/q.ts

```diff
@@ -0,0 +1,157 @@
+/* eslint-disable no-bitwise */
+/*
+ * Copyright © 2022 Lisk Foundation
+ *
+ * See the LICENSE file at the top-level directory of this distribution
+ * for licensing information.
+ *
+ * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
+ * no part of this software, including this file, may be copied, modified,
+ * propagated, or distributed except according to the terms contained in the
+ * LICENSE file.
+ *
+ * Removal or modification of this copyright notice is prohibited.
+ */
+
+import { ONE, Q_OPERATION, TWO, ZERO } from './constants';
+
+const numberToQ = (val: number, base: bigint): bigint => {
+	if (Number.isInteger(val)) {
+		return BigInt(val) << base;
+	}
+	const [intStr, fractionalStr] = val.toString().split('.');
+	return (
+		(BigInt(intStr) << base) +
+		(BigInt(fractionalStr) << base) / BigInt(10) ** BigInt(fractionalStr.length)
+	);
+};
+
+export const q = (val: number | bigint | Buffer, base: number | bigint): Q =>
+	Q.fromValue(val, base);
+
+export class Q {
+	private readonly _base: bigint;
+	private readonly _val: bigint;
+
+	public constructor(val: bigint, base: number | bigint) {
+		this._val = val;
+		this._base = BigInt(base);
+	}
+
+	public static fromValue(val: number | bigint | Buffer | Q, base: number | bigint) {
+		if (val instanceof Q) {
+			return new Q(val._val, val._base);
+		}
+		const bigintBase = BigInt(base);
+		if (typeof val === 'number') {
+			return new Q(numberToQ(val, bigintBase), bigintBase);
+		}
+		if (typeof val === 'bigint') {
+			return new Q(val * TWO ** bigintBase, bigintBase);
+		}
+		return new Q(BigInt(`0x${val.toString('hex')}`), bigintBase);
+	}
+
+	public add(n: Q) {
+		this._checkBase(n);
+
+		return new Q(this._val + n._val, this._base);
+	}
+
+	public sub(n: Q) {
+		this._checkBase(n);
+		if (this._val < n._val) {
+			throw new Error('Invalid input, output cannot be negative.');
+		}
+
+		return new Q(this._val - n._val, this._base);
+	}
+
+	public mul(n: Q) {
+		this._checkBase(n);
+
+		return new Q(this._roundDown(this._val * n._val), this._base);
```

Comment:
While using `_roundDown()` here is mathematically equivalent as what we need to do, semantically it is very confusing.

To avoid confusion, better use:

```typescript
return new Q(this._val * n._val >> this._base, this._base);
```